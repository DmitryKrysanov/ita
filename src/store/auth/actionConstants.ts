export const SIGNIN_SUCCESSFUL = "SIGNIN_SUCCESSFUL";
export const SIGNIN_ERROR = "SIGNIN_ERROR";
export const SIGNOUT = "SIGNOUT";
