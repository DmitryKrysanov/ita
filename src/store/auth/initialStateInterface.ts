export interface InitialStateInterface {
    userReducer: UserInterface
}

export interface UserInterface {
    uid: string | null,
    token: string | null,
    error: null
}