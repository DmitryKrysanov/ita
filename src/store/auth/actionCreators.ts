import * as actionTypes from './actionConstants';
import { SignupInterface, AuthDataInterface } from '@components/App/Auth/Signup/SignupInterface';
import { SIGNIN_SUCCESSFUL, SIGNIN_ERROR, SIGNOUT } from "./actionConstants";
import { Firebase } from "../../components/App/Auth/firebase.config";
import { createUserCoockies, deleteUserCookies } from '../../components/App/Auth/AuthUtils';

export const successSignin = (user: any) => {
    return {
      type: SIGNIN_SUCCESSFUL,
      payload: user
    };
  };
  
  const errorSignin = (err: Error) => {
    return {
      type: SIGNIN_ERROR,
      payload: err
    };
  };
  
  export const signInUser = (email: string, password: string) => {
    return async (dispatch: any) => {
      try {
        const response = await Firebase.auth().signInWithEmailAndPassword(
          email,
          password
        );
  
        if (response.user) {
          console.dir(response.user.getIdTokenResult());
          const token = await response.user
            .getIdTokenResult(true)
            .then(idToken => idToken.token);
          const refreshToken = response.user.refreshToken;
          createUserCoockies(token, refreshToken, response.user.uid);
          dispatch(successSignin({ ...response.user, token }));
        }
      } catch (err) {
        dispatch(errorSignin(err.message));
      }
    };
  };
  export const signOutUser = () => {
    return (dispatch: any) => {
      Firebase.auth()
        .signOut()
        .then(() => {
          deleteUserCookies();
          dispatch({
            type: SIGNOUT
          });
        });
    };
  };

export const saveUserAuthData = (userData: AuthDataInterface) => {
    return {
        type: actionTypes.AUTH_SUCCESSFUL,
        payload: userData
    }
};

export const authSignup = (userData: SignupInterface ) => ({
    type: actionTypes.AUTH_SIGNUP,
    payload: userData
});