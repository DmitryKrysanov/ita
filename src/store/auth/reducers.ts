import { combineReducers } from 'redux'
import { SIGNIN_ERROR, SIGNIN_SUCCESSFUL, SIGNOUT} from './actionConstants'
import {appReducer} from "../reduces";
import { UserInterface } from "./initialStateInterface";
import Cookies from "js-cookie";


let token = Cookies.get('token') || Cookies.get('refreshToken');
const id = Cookies.get('uid');

export const initialState: UserInterface = {
  uid: id || null,
  token: token || null,
  error: null
};

export const userReducer = (state = initialState, action: any) => {
    console.log(action.type);
    switch (action.type) {
        case SIGNIN_SUCCESSFUL:
            return {
                ...state,
                token: action.payload.token || state.token,
                uid: action.payload.uid || state.uid,
                error: null
            };
        case SIGNIN_ERROR:
            return {
                ...state,
                error: action.payload,
            };
        case SIGNOUT:
            return {
                ...state = initialState,
                token: null,
                uid: null
            };
        default:
            return state
    }
};

const rootReducer = combineReducers({
    userReducer,
    appReducer
});

export default rootReducer;