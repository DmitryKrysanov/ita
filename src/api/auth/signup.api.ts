import { SignupInterface, SignupFullDataInterface } from '@components/App/Auth/Signup/SignupInterface';
import { auth, db } from '../../components/App/Auth/firebase.config';
import { signInUser } from '../../store/auth/actionCreators';


export const authSignup = (userData: SignupInterface) => {
  return auth
    .createUserWithEmailAndPassword(userData.email, userData.password)
    .then((data: any) => {console.dir(data);signInUser(userData.email, userData.password); return data.user})
    .catch(() => null);
};

export const authSignupFullData = (fullUserData: SignupFullDataInterface) => {
  console.log('hi');
  console.log(fullUserData);
  return db
    .collection('users')
    .doc(fullUserData.id)
    .set({
      email: fullUserData.email,
      name: fullUserData.name,
      group: fullUserData.group
    })
    .then(() => true)
    .catch(() =>  false);
};