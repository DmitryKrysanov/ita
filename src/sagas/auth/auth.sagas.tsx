import { takeLatest, call, fork, put } from 'redux-saga/effects';
import * as actionTypes from '../../store/auth/actionConstants';
import * as actions from '../../store/auth/actionCreators';
import * as API from '../../api/auth/signup.api';
import { AuthDataInterface, SignupFullDataInterface } from '@components/App/Auth/Signup/SignupInterface';
import { createUserInformation } from '../../components/App/Auth/AuthUtils';

function* getAuth(authData: any) {
  const userToken: string = yield authData.getIdToken()
    .then((token: string) => token );
  const userId: string = authData.uid;
  const refreshToken: string = authData.refreshToken;
  const auth: AuthDataInterface = {
    token: userToken,
    refreshToken: refreshToken,
    uid: userId
  }
  return auth;
}

function* signupWithFullUserData(data: SignupFullDataInterface) {
  const isSuccess: boolean = yield call(API.authSignupFullData, {
    uid: data.uid,
    email: data.email,
    name: data.name,
    group: data.group,
  });
  return isSuccess;
}

function* saveDataInState(userAuthData: any) {
  const storageUserData: AuthDataInterface = {
    token: userAuthData.token,
    uid: userAuthData.uid,
    refreshToken: userAuthData.refreshToken
  }
  yield put(actions.saveUserAuthData(
    storageUserData
  ));
  
  if (!storageUserData.token || !storageUserData.uid || !storageUserData.refreshToken) { return; }
  createUserInformation(storageUserData.token, storageUserData.refreshToken, storageUserData.uid);
}

function* signupUser(action: any) {
  try{
    const user = yield call(API.authSignup, {
      email: action.payload.email,
      password: action.payload.password,
      name: action.payload.name,
      group: action.payload.group,
    });

    if (!user) { return; }

    const userAuthData: AuthDataInterface = yield call(getAuth, user);
    const isSuccess = yield call(signupWithFullUserData, { uid: userAuthData.uid, ...action.payload });

    if(!isSuccess) { return; }

    yield call(saveDataInState, {...userAuthData, error: null, isAuth: true });
  }catch(e){
    // error handle in future
  }
}

function* watchSignupUser() {
  yield takeLatest(actionTypes.AUTH_SIGNUP, signupUser);
}

const authSagas = [
  fork(watchSignupUser)
];

export default authSagas;