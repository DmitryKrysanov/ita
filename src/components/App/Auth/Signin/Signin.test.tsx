import React from "react";
import configureStore, { MockStore } from "redux-mock-store";
import { render, cleanup, fireEvent } from "@testing-library/react";
import { Provider } from "react-redux";
import "@testing-library/jest-dom/extend-expect";
import { initialState } from "../../../../store/auth/reducers";
import SignIn from "./Signin";
import { signInUser } from "../../../../store/auth/actionCreators";

const mockStore = configureStore([]);
afterEach(cleanup);
describe("Connected React-Redux Component", () => {
  let store: MockStore;
  let component: any;
  beforeEach(() => {
    store = mockStore({
      userReducer: { ...initialState },
      appReducer: { projectTitle: "ITA-tools" }
    });
    store.dispatch = jest.fn();
    component = render(
      <Provider store={store}>
        <SignIn />
      </Provider>
    );
  });

  it("should render with given state from Redux store", () => {
    expect(component).toMatchSnapshot();
  });

  it("should render buttons", () => {
    const { getByText } = component;
    expect(getByText("Sign in")).toBeInTheDocument();
    expect(getByText("Sign out")).toBeInTheDocument();
    expect(getByText("Submit")).toBeInTheDocument();
  });

  it("should close the window by clicking on icon", () => {
    const { baseElement } = component;
    const closeIcon = baseElement.querySelector("i");
    expect(closeIcon).toBeInTheDocument();
  });
  it("should sign out by clicking on button", async () => {
    const { baseElement } = component;
    const SignoutBtn = baseElement.querySelector(".signOut");
    fireEvent.click(SignoutBtn);
    expect(store.dispatch).toHaveBeenCalledTimes(1);
  });

  it("should render email and password input", () => {
    const { baseElement } = component;
    const inputs = baseElement.querySelectorAll("input");
    expect(inputs).toHaveLength(2);
  });

  it("should submit form by clicking on button", async () => {
    const { baseElement } = component;
    const submit = baseElement.querySelector('button[type="submit"]');
    const email = baseElement.querySelector('input[type="email"]');
    const password = baseElement.querySelector('input[type="password"]');

    fireEvent.change(email, { target: { value: "softserve@gmail" } });
    fireEvent.change(password, { target: { value: "123serve" } });
    fireEvent.submit(submit);

    expect(email.value).toEqual("softserve@gmail");
    expect(password.value).toEqual("123serve");
    expect.assertions(2);
    try {
      await signInUser("softserve@gmail", "123serve");
    } catch (e) {
      expect(e).toEqual({
        error: "User not found"
      });
    }
  });
});
