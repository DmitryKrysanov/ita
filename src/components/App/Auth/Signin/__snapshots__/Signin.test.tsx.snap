// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connected React-Redux Component should render with given state from Redux store 1`] = `undefined`;
